mir_eval/sonify.py

```diff
@@ -147,34 +154,42 @@ def __interpolator(x):
             return value
         return __interpolator
 
+    # print(f"sonify.time_frequency length {length}, n_times {n_times}, gram shape {gram.shape}, times shape {times.shape}")
+
     # Threshold the tfgram to remove non-positive values
     gram = np.maximum(gram, 0)
 
     # Pre-allocate output signal
     output = np.zeros(length)
     time_centers = np.mean(times, axis=1) * float(fs)
 
+    # Not really a true spectral energy, just a summation for optimisation.
+    spectral_sums = np.sum(gram, axis = 1)
     for n, frequency in enumerate(frequencies):
+        if spectral_sums[n] < 0.01: # TODO set threshold intelligently.
+            continue
         # Get a waveform of length samples at this frequency
         wave = _fast_synthesize(frequency)
+        # print(f"Freq {n} {frequency}Hz energy {spectral_sums[n]}, length of wave {len(wave)}")
 
         # Interpolate the values in gram over the time grid
         if len(time_centers) > 1:
             gram_interpolator = interp1d(
-                time_centers, gram[n, :],
+                time_centers, gram[n, :-1],
```

Comment:
why are we now off by one here?

Reply:
When converting `times` from boundaries to intervals with `utils.boundaries_to_intervals()`, this will result in a reduction in length of `times` by 1, which needs reconciling with the length of `gram`. Historically, the bug in using `length` to adjust the intervals left a very long last interval, which covered up the difference in length.

Reply:
Sorry, still not following this.  But I think it would be clearer if instead of `:-1`, it explicitly matched the length of `time_centers`, which should be exactly `n_times` right?  Explicit > implicit and all that.

Reply:
Fair comment, I've now done that.